metadata/metadata.c

```diff
@@ -148,16 +160,75 @@ Allocator *create_allocator(VirtualDisk__ErasureCodeProfile ec_profile)
     return result;
 }
 
-Metadata *new_metadata()
+HyperwarpMetadata *new_metadata(const char *backend, VirtualDisk__ErasureCodeProfile ec_profile)
 {
-    Metadata *metadata = (Metadata *)malloc(sizeof(Metadata));
-    metadata__init(metadata);
+    HyperwarpMetadata *metadata = (HyperwarpMetadata *)calloc(1, sizeof(HyperwarpMetadata));
+    metadata->metadata = (Metadata*)calloc(1, sizeof(Metadata));
+    metadata__init(metadata->metadata);
+    metadata->metadata->ec_profile = ec_profile;
+    metadata->allocator = create_allocator(ec_profile);
+
+    if (use_metadata_storage_backend(metadata, backend) != 0) {
+        fprintf(stderr, "Could not initialize the selected storage backend");
+        free(metadata);
+        return NULL;
+    }
+
+    if (metadata_backend_initialize(metadata) != 0) {
+        fprintf(stderr, "Could not initialize the selected storage backend\n");
+        free(metadata);
+        return NULL;
+    }
 
     return metadata;
 }
 
-void _add_physical_disk(Metadata *metadata, PhysicalDisk *disk)
+HyperwarpMetadata *load_metadata(const char *backend) {
+    HyperwarpMetadata *metadata = (HyperwarpMetadata *)calloc(1, sizeof(HyperwarpMetadata));
+
+    if (use_metadata_storage_backend(metadata, backend) != 0) {
+        fprintf(stderr, "Could not initialize the selected storage backend");
+        free(metadata);
+        return NULL;
+    }
+
+    if (metadata_backend_initialize(metadata) != 0) {
+        fprintf(stderr, "Could not initialize the selected storage backend\n");
+        free(metadata);
+        return NULL;
+    }
+
+    metadata->metadata = metadata->backend_context->backend->load();
+
+    if (metadata->metadata != NULL) {
+        metadata->allocator = create_allocator(metadata->metadata->ec_profile);
+        return metadata;
+    } else {
+        metadata_close(metadata);
+        return NULL;
+    }
+}
+
+void metadata_close(HyperwarpMetadata *metadata) {
+
+    metadata__free_unpacked(metadata->metadata, NULL);
+    metadata->metadata = NULL;
+
+    if (metadata->backend_context->backend_lib_handler != NULL) {
+        metadata->backend_context->backend->finalize();
+        dlclose(metadata->backend_context->backend_lib_handler);
+    }
+
+    if (metadata->allocator != NULL) {
+        free(metadata->allocator);
+    }
+
+    free(metadata);
+}
+
+void _add_physical_disk(HyperwarpMetadata *hwmetadata, PhysicalDisk *disk)
 {
+    Metadata *metadata = hwmetadata->metadata;
```

Comment:
this change doesn't seem necessary right now. You could just continue to pass the proto Metadata struct and it would be fine.